File: apps/MitSMK/frontend/src/stores/artworks.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { api } from '../services/api'
import type { Artwork, ArtworkSearchResponse } from '../types/api'

export const useArtworksStore = defineStore('artworks', () => {
  const searchResults = ref<Artwork[]>([])
  const currentArtwork = ref<Artwork | null>(null)
  const isLoading = ref(false)
  const searchQuery = ref('')
  const totalResults = ref(0)
  const currentOffset = ref(0)
  const limit = ref(20)

  const searchArtworks = async (query: string, offset: number = 0) => {
    if (!query.trim()) return
    
    isLoading.value = true
    searchQuery.value = query
    currentOffset.value = offset
    
    try {
      const response = await api.get<ArtworkSearchResponse>('/artworks/search', {
        params: {
          q: query,
          offset,
          limit: limit.value
        }
      })
      
      if (offset === 0) {
        searchResults.value = response.data.artworks
      } else {
        searchResults.value = [...searchResults.value, ...response.data.artworks]
      }
      
      totalResults.value = response.data.total
    } catch (error) {
      console.error('Error searching artworks:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const getArtwork = async (objectNumber: string) => {
    isLoading.value = true
    
    try {
      const response = await api.get<Artwork>(`/artworks/${objectNumber}`)
      currentArtwork.value = response.data
      return response.data
    } catch (error) {
      console.error('Error fetching artwork:', error)
      throw error
    } finally {
      isLoading.value = false
    }
  }

  const clearSearch = () => {
    searchResults.value = []
    searchQuery.value = ''
    totalResults.value = 0
    currentOffset.value = 0
  }

  const loadMore = async () => {
    if (searchQuery.value && !isLoading.value) {
      await searchArtworks(searchQuery.value, currentOffset.value + limit.value)
    }
  }

  return {
    searchResults,
    currentArtwork,
    isLoading,
    searchQuery,
    totalResults,
    currentOffset,
    limit,
    searchArtworks,
    getArtwork,
    clearSearch,
    loadMore
  }
})

File: apps/MitSMK/backend/app/routers/artworks.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from typing import List

from ..database import get_db
from ..schemas import Artwork, ArtworkSearchResponse
from ..smk_service import SMKService

router = APIRouter()
smk_service = SMKService()

@router.get("/search", response_model=ArtworkSearchResponse)
async def search_artworks(
    q: str = Query(..., description="Search query"),
    offset: int = Query(0, ge=0, description="Offset for pagination"),
    limit: int = Query(20, ge=1, le=100, description="Number of results to return"),
    db: Session = Depends(get_db)
):
    """Search for artworks using SMK API"""
    if not q.strip():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Search query cannot be empty"
        )
    
    try:
        # Search and cache artworks
        artworks = await smk_service.search_and_cache_artworks(q, offset, limit, db)
        
        # Get total count from SMK API for pagination
        search_result = await smk_service.search_artworks(q, 0, 1)
        total = search_result.get("found", len(artworks))
        
        return ArtworkSearchResponse(
            artworks=artworks,
            total=total,
            offset=offset,
            limit=limit
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error searching artworks: {str(e)}"
        )

@router.get("/{object_number}", response_model=Artwork)
async def get_artwork(
    object_number: str,
    db: Session = Depends(get_db)
):
    """Get detailed information about a specific artwork"""
    try:
        artwork = await smk_service.get_artwork_by_object_number(object_number, db)
        
        if not artwork:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Artwork not found"
            )
        
        return artwork
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error fetching artwork: {str(e)}"
        )

